Fix for double backwards tests

ifedan opened the pull request:
If none of the outputs require_grad, we don't actually check gradgrad, instead we will check that their numerical gradients are 0.

ifedan committed:
Fix for double backwards tests

ezyang commented:
Hmm... I'm a little perplexed how this code ever worked. In your new codepath, you index based on position in `func_out`; in the old codepath, we index based on position in `output`. But non-differentiable outputs might be filtered out. So the old codepath sounds wrong, unless we're lucky and non-differentiable outputs only ever happen in the suffix. Do you know if this is true?

You have no test code changes in this diff, which makes it hard to tell if this actually worked or not.

SsnL commented:
> If none of the outputs require_grad, we don't actually check gradgrad, instead we will check that their numerical gradients are 0.

This is only valid if all inputs to the function require grad!

gchanan commented on torch/autograd/gradcheck.py:
```diff
@@ -225,7 +225,18 @@ def fail_test(msg):
+                return _as_tuple(func(*input))[i]
+            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
+            for n in numerical:
+                if not n.sum().item() == 0:
```
this doesn't check whether the numerical gradient is 0...

ifedan replied:
Fixed

gchanan commented:
how about a test function that would pass without this change and fail with it?

ifedan committed:
Merge branch 'master' into fix_gradgrad

ifedan committed:
Merge branch 'master' into fix_gradgrad

ifedan committed:
changed the way to verify that numerical gradient is 0

ifedan commented:
> > If none of the outputs require_grad, we don't actually check gradgrad, instead we will check that their numerical gradients are 0.
> 
> This is only valid if all inputs to the function require grad!

Yes, we filter out inputs: input_args = tuple(x for x in input_args if isinstance(x, torch.Tensor) and x.requires_grad)

gchanan commented:
@SsnL I wonder what the right thing to do here is.  Maybe we can just throw an exception if there are no differentiable outputs and all of the inputs don't require grad because we can't really check that.   Although what do we do about integral inputs (which can never have requires_grad).  And I don't know if things call this today with those properties that would start failing.

gchanan commented:
and there's non-differentiable outputs for functions that have some differentiable outputs as well...(although we know they are fine if the non-differentiable outputs are integral, which is the usual case with indices).

ifedan committed:
Added test for gradcheck and output with requires_grad=False

ifedan commented:
> how about a test function that would pass without this change and fail with it?

Added test

ifedan commented:
> Hmm... I'm a little perplexed how this code ever worked. In your new codepath, you index based on position in `func_out`; in the old codepath, we index based on position in `output`. But non-differentiable outputs might be filtered out. So the old codepath sounds wrong, unless we're lucky and non-differentiable outputs only ever happen in the suffix. Do you know if this is true?
> 
> You have no test code changes in this diff, which makes it hard to tell if this actually worked or not.

Previously when there were no differentiable outputs then checkgrad will do nothing, so I added check for numerical gradients.
Also I added test for this section.

gchanan commented on test/test_autograd.py:
```diff
@@ -2285,6 +2285,15 @@ def rfft_irfft(x):
         _test_complex((1, 2, 3, 4, 2), 2)
         _test_complex((2, 1, 3, 4, 3, 2), 3)
 
+    def test_gradcheck(self):
```
please name this better -- there are ton of things that we could check about gradcheck and this is one single example.

ifedan replied:
Fixed

gchanan commented on torch/autograd/gradcheck.py:
```diff
@@ -225,7 +225,18 @@ def fail_test(msg):
+    output = _differentiable_outputs(func_out)
+
+    if not output:
+        for i, o in enumerate(func_out):
```
what about to @SsnL's point that this is only valid if all of the inputs require_grad?

SsnL replied:
I think @ifedan should filter out inputs that don't require grad in lines 233~235.

gchanan replied:
I'm a little concerned that would also miss cases we want to check.  I'd probably just start by asserting that if there are no differentiable outputs and at least one requires_grad input, then all the inputs requires_grad.  Then we could loosen the requirements later if there are cases that run afoul of that check.

ifedan replied:
Added check that if any input requires grad than all the inputs must require grad

SsnL commented on torch/autograd/gradcheck.py:
```diff
@@ -225,7 +225,18 @@ def fail_test(msg):
+            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
+            for n in numerical:
+                if len(torch.nonzero(n)) > 0:
+                    fail_test('Numerical gradient for function expected to be zero')
```
You should return here

SsnL replied:
Could you also add a return in line 207? Thanks!

ifedan replied:
Fixed

ifedan committed:
changes based on PR review

gchanan commented on torch/autograd/gradcheck.py:
```diff
@@ -206,9 +206,12 @@ def fail_test(msg):
     if any(t.is_sparse for t in tupled_inputs if isinstance(t, torch.Tensor)) and not check_sparse_nnz:
         fail_test('gradcheck expects all tensor inputs '
                   'are dense when check_sparse_nnz is set to False.')
+        return False
```
you should do "return fail_test" rather than separately returning False.

gchanan commented on torch/autograd/gradcheck.py:
```diff
@@ -219,13 +222,31 @@ def fail_test(msg):
+            numerical = get_numerical_jacobian(fn, tupled_inputs, eps=eps)
+            for n in numerical:
+                if len(torch.nonzero(n)) > 0:
+                    fail_test('Numerical gradient for function expected to be zero')
```
same here.

gchanan reviewed (approved):
approved with comments.

ifedan committed:
changes based on PR review

ifedan committed:
Merge branch 'master' into fix_gradgrad

ezyang commented:
This failed lint
